wordpress-mobile/WordPress-iOS
# Reader Comments: Highlight the cell when it's focused or navigated to

Refs #17476, pbArwn-2ks-p2#comment-4438
Depends on #17635

![](https://user-images.githubusercontent.com/1299411/141975576-3abbe18f-df47-48df-b144-deb7871b64b9.png)

As titled, this implements the new highlighted style as discussed in `pbArwn-2ks-p2#comment-4438`. A couple of notes on the behavior:

- There are two components that can be highlighted: the background, and the Reply button.
  - When navigating to a specific comment (i.e. from Notification comments), only the background is highlighted.
  - When tapping on the Reply button, both the background and the Reply button are highlighted.
- With a cell already highlighted, tapping anywhere on the screen un-highlights the cell, hides the keyboard, and updates the Reply text view to target the post.
- When a cell is already highlighted, tapping the Reply button on a different cell will un-highlight the first one, and highlights the new one.
- When tapping on an already highlighted Reply button, it will un-highlight the cell. The Reply text view should be updated to reply to the post.

I also noticed some legacy issues that have preexisted and decided not to tackle them for now:

1. With a cell highlighted and the keyboard is shown, tapping anywhere on the table view (except interactable elements like the Reply button, etc.) should un-highlight the cell and hide the keyboard. However, tapping on a `UITextView` will hide the keyboard, but will NOT un-highlight the cell. 
    - Notice that the Reply text view placeholder is still "Reply to <author>". This is also true with the previous implementation. I'm not sure why it behaved like this, but I think this is not a major issue and I can take a look next week if this needs to be fixed in 18.9.
2. After submitting a reply, there's this "makeshift" comment that's inserted into the table, and then updated after the upload process completes:
  ![Simulator Screen Shot - iPhone 12 - 2021-12-09 at 19 25 33](https://user-images.githubusercontent.com/1299411/145405709-c810d9fe-b0b1-454d-b205-c6f1a9a06680.png)
  This is because a local comment is created in the `CommentService`, and saved before hitting the remote, and this informs the results controller which in turn updates the table view. I'm holding off from fixing this for now as it could go quite deep into the hierarchy logic and the comments service... but I can create an issue later for this one.

## To test

- Enable the `newCommentThread` feature flag.
- Go to Notifications, and select any comment notification.
- Tap on the header to go into the comment threads.
- 🔍 Verify that the comment background is highlighted, and the Reply button is NOT highlighted.
- 🔍 Verify that the Reply text view placeholder shows "Reply to post".
- Tap a Reply button on a different row.
- 🔍 Verify that the previously highlighted cell is now unhighlighted.
- 🔍 Verify that on the selected row, the background is highlighted and the Reply button is highlighted.
- 🔍 Verify that the Reply text view placeholder is updated to show the new author.
- Try sending a reply.
- 🔍 Verify that a new comment reply is created, and the hierarchy is correct.
- Tap on the highlighted Reply button.
- 🔍 Verify that the row is un-highlighted: background is NOT highlighted, Reply button is NOT highlighted.
- 🔍 Verify that the Reply text view shows "Reply to post".
- Try sending a reply.
- 🔍 Verify that a new post reply is created.

## Regression Notes
1. Potential unintended areas of impact
N/A. Feature is hidden behind a flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A. Feature is hidden behind a flag.

3. What automated tests I added (or what prevented me from doing so)
N/A. Feature is hidden behind a flag.

PR submission checklist:
